Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,322 @@
+lemma utility_winner  (H: i = winner b) : utility b i = a.v i - secondprice b:= by
+  rw [utility]; simp only [ite_true, H]
+
+/-- If `i` is not the winner, then their utility is 0. -/
+lemma utility_loser (i: a.I) (H : i≠ winner b) : utility b i = 0 := by
+  rw [utility]; simp only [ite_false, H]
+
+/-- A strategy is dominant if bidding `bi` ensures that
+`i`'s utility is maximized relative to any other bids `b'` where `b i = bi`. -/
+def dominant (i : a.I) (bi : ℝ) : Prop :=
+  ∀ b b': a.I → ℝ , (b i = bi) → (∀ j: a.I, j≠ i→ b j = b' j)
+   →  utility  b i ≥ utility b' i
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma gt_wins (i : a.I) (H: ∀ j , i ≠ j →  b i > b j) : i = winner b := by
+   have HH : ∀ j, i = j ↔  b j = maxb b:= by
+      have imax : b i = maxb b := by
+         have H1 : b i ≤  maxb b := by
+            apply Finset.le_sup'
+            simp only [Finset.mem_univ]
+         have H2 : maxb b ≤ b i := by
+            apply Finset.sup'_le
+            intro j _
+            by_cases hji : i=j
+            · rw [hji]
+            · exact le_of_lt (H j hji)
+         simp only [maxb]
+         exact Real.partialOrder.proof_4 (b i) (Finset.sup' Finset.univ maxb.proof_1 b) H1 H2
+      intro j
+      constructor
+      ·  intro hji
+         rw [<-hji]
+         exact imax
+      ·  intro hbj
+         by_contra hji
+         have hji' := H j (by rw [ne_eq];exact hji)
+         aesop
+   rw [HH]
+   rw [<-winner_take_max]
+
+/-- The bid of the winner is always greater than or equal to the bids of all other participants.-/
+lemma b_winner_max (H: i = winner b) : ∀ j: a.I, b i ≥ b j := by
+  intro j
+  have H_max: b (winner b) = maxb b := winner_take_max b
+  rw [<-H] at H_max
+  have H_sup: b j ≤ maxb b := by
+    apply Finset.le_sup'; simp only [Finset.mem_univ]
+  rw [<-H_max] at H_sup
+  exact H_sup
+
+
+/-- The bid of the winner is always greater than or equal to the second highest bid. -/
+lemma b_winner (H: i = winner b) : b i ≥ secondprice b := by
+  have Hmax := winner_take_max b
+  rw [<-H] at Hmax
+  rw [Hmax]
+  apply Finset.sup'_le
+  apply delete_i_nonempty
+  intro j _
+  apply Finset.le_sup'
+  simp only [Finset.mem_erase, Finset.mem_univ]
+
+/-- If `i` is not the winner, then the highest bid excluding `i` is equal to the highest bid. -/
+lemma b_loser_max (H: i ≠ winner b) : B b i = maxb b := by
+  have H1: B b i ≤ maxb b := by
+    apply Finset.sup'_le
+    intro i _
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ]
+  have H2: maxb b ≤ B b i := by
+    rw [<-winner_take_max b]
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ, Finset.mem_erase, and_true]
+    exact (Ne.symm H)
+  exact Real.partialOrder.proof_4 (B b i) (Finset.sup' Finset.univ maxb.proof_1 b) H1 H2
+
+
+
+/-- Utility is non-negative if the bid equals the valuation. -/
+lemma utility_nneg (i: a.I) : (b i = a.v i) → utility b i ≥ 0 := by
+  intro H
+  by_cases H2: i = winner b
```

Comment:
After a `by_cases`, it's best to create a `·` for each goal: it makes it easier to read without lean running, but also it means that if someone breaks your proof, they know which goal they broke.